src/ORM/UniqueKey/UniqueKeyService.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace SilverStripe\ORM\UniqueKey;
+
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * Class UniqueKeyService
+ *
+ * Generate a unique key for data object
+ *
+ * recommended use:
+ * - when you need unique key for caching purposes
+ * - when you need unique id on the front end (for example JavaScript needs to target specific element)
+ *
+ * @package SilverStripe\ORM\UniqueKey
+ */
+class UniqueKeyService implements UniqueKeyInterface
+{
+    use Injectable;
+
+    public function generateKey(DataObject $object, array $extraKeys = []): string
+    {
+        if (!$object->isInDB()) {
+            return '';
```

Comment:
I wonder if an unsaved object should return a random number instead? The behaviour would in effect be that if the unique key of and unsaved object turned up in a cache, it would be uncacheable.

This seems safer than the current implementation, which would mean that *all* unsaved objects are saved to the *same* cache key, which feels like a future-PII-leak. ;-)